### PR TITLE
[cbc] Add new port

### LINKED ISSUE #10425: [New Port Request] cbc
Library name: cbc

Library description: COIN-OR Branch-and-Cut solver

Source repository URL: https://github.com/coin-or/Cbc

Project homepage (if different from the source repository):

Anything else that is useful to know when adding (such as optional features the library may have that should be included): ortools depends on this.

### PR DESCRIPTION
**Pull request by NancyLi1013**
This port is one of or-tools dependencies.

No feature need to test.
Fixes #10425

### CODE REVIEW

**Comment by NancyLi1013** [comment]
It needs to depend on cgl #7810.

**Comment by NancyLi1013** [comment]
This port failed on all triplets, due to the following reason:

> CMake Error at C:/vsts/_work/1/s/installed/x64-uwp/share/cgl/CglTargets.cmake:74 (message):
>   The imported target "Coin::Cgl" references the file
> 
>      "C:/vsts/_work/1/s/packages/cgl_x64-uwp/lib/Cgl.lib"
> 
>   but this file does not exist.  Possible reasons include:
> 
>   * The file was deleted, renamed, or moved to another location.
> 
>   * An install or uninstall procedure did not complete successfully.
> 
>   * The installation package was faulty and contained
> 
>      "C:/vsts/_work/1/s/installed/x64-uwp/share/cgl/CglTargets.cmake"
> 
>   but not all the files it references.
> 
> Call Stack (most recent call first):
>   C:/vsts/_work/1/s/installed/x64-uwp/share/cgl/CglConfig.cmake:28 (include)
>   C:/vsts/_work/1/s/scripts/buildsystems/vcpkg.cmake:256 (_find_package)
>   CMakeLists.txt:44 (find_package)
> 
> 
> -- Configuring incomplete, errors occurred!

I wondered why it cannot find cgl. I have merged cgl to this branch, but it still build failed.

File: ports/cbc/Config.cmake.in [hunk 1]
```diff
@@ -0,0 +1,5 @@
+
+@PACKAGE_INIT@
+
```
Comment (vicroms) [inline ports/cbc/Config.cmake.in]: Should add these lines here, in order to propagate dependencies to libraries linking to `cbc`:

```
include(CMakeFindDependencyMacro)
find_dependency(CoinUtils REQUIRED)
find_dependency(Osi REQUIRED)
find_dependency(Cgl REQUIRED)
find_dependency(Clp REQUIRED)
```

File: ports/cbc/portfile.cmake [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+file(COPY ${CMAKE_CURRENT_LIST_DIR}/CMakeLists.txt DESTINATION ${SOURCE_PATH})
+file(COPY ${CMAKE_CURRENT_LIST_DIR}/Config.cmake.in DESTINATION ${SOURCE_PATH})
+
+vcpkg_configure_cmake(
+    SOURCE_PATH ${SOURCE_PATH}
+    PREFER_NINJA
+)
+
+vcpkg_install_cmake()
+
```
Comment (vicroms) [inline ports/cbc/portfile.cmake]: Needs a call to `vcpkg_fixup_cmake_targets()` after configuring and installing.

**Review by vicroms — CHANGES_REQUESTED** [review]

File: ports/cbc/portfile.cmake (hunk 2 above)
Comment (NancyLi1013) [reply ports/cbc/portfile.cmake]: @vicroms  Thanks for your review and suggestions. I have updated.

**Comment by NancyLi1013** [comment]
The failures are the same as before.

**Comment by PhoebeHui** [comment]
See failures "C:/vsts/_work/1/s/packages/cgl_x64-uwp/lib/Cgl.lib", please check cgl, it shouldn't reference anything in packages directory.

**Comment by NancyLi1013** [comment]
Hi @dan-shaw maybe you closed this PR as you misread this as an solved issue.
So I reopen it.

**Review by PhoebeHui — APPROVED** [review]
LGTM

File: ports/cbc/fix-c1083-error.patch [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+diff --git a/Cbc/src/CbcCutGenerator.cpp b/Cbc/src/CbcCutGenerator.cpp
+index 995315a..67db8b1 100644
+--- a/Cbc/src/CbcCutGenerator.cpp
++++ b/Cbc/src/CbcCutGenerator.cpp
+@@ -26,7 +26,7 @@
+ #include "CbcMessage.hpp"
+ #include "CbcCutGenerator.hpp"
+ #include "CbcBranchDynamic.hpp"
+-#include "CglProbing.hpp"
++#include "CglProbing/CglProbing.hpp"
```
Comment (JackBoosY) [inline ports/cbc/fix-c1083-error.patch]: Can you modify these changes to add these paths to _INCLUDE_PATH_ in CMakeLists.txt ?

**Review by JackBoosY — CHANGES_REQUESTED** [review]

**Comment by NancyLi1013** [comment]
Add these codes to CMakeLists.txt
```
set(CGL_INCLUDE_DIR "include/Coin/CglProbing" 
                     "include/Coin/CglPreProcess" 
                     "include/Coin/CglClique"
                     "include/Coin/CglFlowCover"
                     "include/Coin/CglGomory"
                     "include/Coin/CglKnapsackCover"
                     "include/Coin/CglMixedIntegerRounding2"
                     "include/Coin/CglOddHole"
                     "include/Coin/CglRedSplit"
                     "include/Coin/CglRedSplit2"                  
                     "include/Coin/CglTwomir"
                     "include/Coin/CglDuplicateRow"
                     "include/Coin/CglLandP"
                     "include/Coin/CglGMI"
                     "include/Coin/CglZeroHalf"
                     "include/Coin/CglResidualCapacity"
                     )
list(APPEND RELATIVE_INSTALL_INCLUDE_DIR ${CGL_INCLUDE_DIR})

```
Print out the value of `RELATIVE_INSTALL_INCLUDE_DIR`:

`RELATIVE_INSTALL_INCLUDE_DIR:include/Coin;include/Coin/CglProbing;include/Coin/CglPreProcess;include/Coin/CglClique;include/Coin/CglFlowCover;include/Coin/CglGomory;include/Coin/CglKnapsackCover;include/Coin/CglMixedIntegerRounding2;include/Coin/CglOddHole;include/Coin/CglRedSplit;include/Coin/CglRedSplit2;include/Coin/CglTwomir;include/Coin/CglDuplicateRow;include/Coin/CglLandP;include/Coin/CglGMI;include/Coin/CglZeroHalf;include/Coin/CglResidualCapacity`

The error like this:
```
CMake Error in CMakeLists.txt:
  Target "Cbc" INTERFACE_INCLUDE_DIRECTORIES property contains path:

    "F:/7850/vcpkg/buildtrees/cbc/src/0978144e9f-840ab9ffcc/${_IMPORT_PREFIX}/include/Coin/CglProbing"

  which is prefixed in the source directory.
```
Could anyone help give some suggestion about this?

**Comment by JackBoosY** [comment]
Temporary close this PR, Please reopen it when continue this work.

**Comment by timosachsenberg** [comment]
@NancyLi1013 has there been any update or progress?

**Pull request closed without merging**